kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -425,240 +293,345 @@ private void writeApiVersionsResponse(ChannelHandlerContext ctx, DecodedRequestF
+        return clientCtx().channel().localAddress();
     }
 
+    /**
+     * Accessor exposing the authorizedId to a {@link NetFilter}.
+     * <p>Called by the {@link #netFilter}.</p>
+     * @return The authorized id, or null.
+     * @throws IllegalStateException if {@link #proxyChannelStateMachine} is not {@link SelectingServer}.
+     */
     @Override
     public String authorizedId() {
+        proxyChannelStateMachine.assertIsSelectingServer("NetFilter invoked NetFilterContext accessor outside SelectingServer state");
         return authentication != null ? authentication.authorizationId() : null;
     }
 
+    /**
+     * Accessor exposing the name of the client library to a {@link NetFilter}.
+     * <p>Called by the {@link #netFilter}.</p>
+     * @return The name of the client library, or null.
+     * @throws IllegalStateException if {@link #proxyChannelStateMachine} is not {@link SelectingServer}.
+     */
     @Override
     public String clientSoftwareName() {
-        return clientSoftwareName;
+        if (proxyChannelStateMachine.state() instanceof SelectingServer selectingServer) {
+            return selectingServer.clientSoftwareName();
+        }
+        else {
+            throw new IllegalStateException();
+        }
     }
 
+    /**
+     * Accessor exposing the version of the client library to a {@link NetFilter}.
+     * <p>Called by the {@link #netFilter}.</p>
+     * @return The version of the client library, or null.
+     * @throws IllegalStateException if {@link #proxyChannelStateMachine} is not {@link SelectingServer}.
+     */
     @Override
     public String clientSoftwareVersion() {
-        return clientSoftwareVersion;
+        if (proxyChannelStateMachine.state() instanceof SelectingServer selectingServer) {
+            return selectingServer.clientSoftwareVersion();
+        }
+        else {
+            throw new IllegalStateException();
+        }
     }
 
+    /**
+     * Accessor exposing the SNI host name to a {@link NetFilter}.
+     * <p>Called by the {@link #netFilter}.</p>
+     * @return The SNI host name, or null.
+     * @throws IllegalStateException if {@link #proxyChannelStateMachine} is not {@link SelectingServer}.
+     */
     @Override
     public String sniHostname() {
+        proxyChannelStateMachine.assertIsSelectingServer("NetFilter invoked NetFilterContext accessor outside SelectingServer state");
         return sniHostname;
     }
 
+    /**
+     * Initiates the connection to a server.
+     * Changes {@link #proxyChannelStateMachine} from {@link SelectingServer} to {@link Connecting}
+     * Initializes the {@code backendHandler} and configures its pipeline
+     * with the given {@code filters}.
+     * <p>Called by the {@link #netFilter}.</p>
+     * @param remote upstream broker target
+     * @param filters The protocol filters
+     */
```

Comment:
I wrote a lot of these javadocs before this PR took its eventual form. We should review them because I don't think they're all still correct.